File: Analisis/make_Higgs_hists.cpp
#include <podio/Frame.h>
#include <podio/ROOTReader.h>
#include <edm4hep/ReconstructedParticleCollection.h>
#include <edm4hep/RecoMCParticleLinkCollection.h>
#include <edm4hep/utils/kinematics.h>
#include <TFile.h>
#include <TH1D.h>
#include <iostream>
#include <memory>
#include <string>
#include <cmath>

/**
 * Analysis executable to produce histograms for H -> gamma gamma candidates.
 * Produces: M(gg) post-fit, M(gg) pre-fit, pT and eta of the Higgs and the two photons.
 */
int main(int argc, char* argv[]) {
    if (argc < 3) {
        std::cout << "Usage: ./make_Higgs_hists <input.root> <collection_name> [output.root]" << std::endl;
        std::cout << "  collection_name: typically 'Higgs_New'" << std::endl;
        return 1;
    }
    std::string inputfile  = argv[1];
    std::string colName    = argv[2];
    std::string outputfile = (argc >= 4) ? argv[3] : "Higgs_histograms_output.root";

    podio::ROOTReader reader;
    try {
        reader.openFile(inputfile);
    } catch (const std::exception& e) {
        std::cerr << "Error opening file: " << e.what() << std::endl;
        return 1;
    }

    auto hfile = std::make_unique<TFile>(outputfile.c_str(), "RECREATE");

    // --- Masa M(gg) ---
    auto h_mass     = new TH1D("h_mass",     "Post-fit M(#gamma#gamma);M_{H} [GeV];Entries",                  100, 110.0, 140.0);
    auto h_prefit   = new TH1D("h_prefit",   "Pre-fit M(#gamma#gamma) (Daughter Sum);M_{#gamma#gamma} [GeV];Entries", 100,  90.0, 160.0);
    auto h_gen_mass = new TH1D("h_gen_mass", "Generated (MC Truth) Mass;M_{H} [GeV];Entries",                 100, 110.0, 140.0);
    auto h_res      = new TH1D("h_res",      "Mass Resolution;M_{reco} - M_{gen} [GeV];Entries",              100, -15.0,  15.0);

    // --- pT del Higgs ---
    auto h_pt       = new TH1D("h_pt",       "Higgs Transverse Momentum;p_{T}^{H} [GeV];Entries",             100,   0.0, 200.0);
    // --- pT de cada foton hijo ---
    auto h_gamma_pt = new TH1D("h_gamma_pt", "Photon Transverse Momentum;p_{T}^{#gamma} [GeV];Entries",       100,   0.0, 150.0);

    // --- Pseudorapidez del Higgs ---
    auto h_eta       = new TH1D("h_eta",      "Higgs Pseudorapidity;#eta^{H};Entries",                        100,  -5.0,   5.0);
    // --- Pseudorapidez de cada foton hijo ---
    auto h_gamma_eta = new TH1D("h_gamma_eta","Photon Pseudorapidity;#eta^{#gamma};Entries",                   100,  -5.0,   5.0);

    unsigned int nEvents = reader.getEntries("events");
    std::cout << "[INFO] Processing " << nEvents << " events from collection '" << colName << "'..." << std::endl;

    for (unsigned int i = 0; i < nEvents; ++i) {
        auto event = podio::Frame(reader.readEntry("events", i));
        try {
            const auto& particles = event.get<edm4hep::ReconstructedParticleCollection>(colName);

            const edm4hep::RecoMCParticleLinkCollection* associations = nullptr;
            try {
                associations = &event.get<edm4hep::RecoMCParticleLinkCollection>("RecoMCTruthLink");
            } catch (...) {}

            for (const auto& p : particles) {

                // --- 1. Masa post-fit ---
                double recoMass = p.getMass();
                h_mass->Fill(recoMass);

                // --- 2. pT y eta del Higgs ---
                double px  = p.getMomentum().x;
                double py  = p.getMomentum().y;
                double pz  = p.getMomentum().z;
                double pt  = std::sqrt(px*px + py*py);
                double eta = (pt > 0) ? std::asinh(pz / pt) : 0.0;
                h_pt->Fill(pt);
                h_eta->Fill(eta);

                // --- 3. Masa pre-fit y cinematica de los fotones hijos ---
                auto daughters = p.getParticles();
                if (daughters.size() >= 2) {
                    auto p1 = edm4hep::utils::p4(daughters[0], edm4hep::utils::UseEnergy);
                    auto p2 = edm4hep::utils::p4(daughters[1], edm4hep::utils::UseEnergy);
                    double prefitMass = (p1 + p2).M();
                    h_prefit->Fill(prefitMass);

                    // pT y eta de cada foton
                    for (int d = 0; d < 2; ++d) {
                        double gpx  = daughters[d].getMomentum().x;
                        double gpy  = daughters[d].getMomentum().y;
                        double gpz  = daughters[d].getMomentum().z;
                        double gpt  = std::sqrt(gpx*gpx + gpy*gpy);
                        double geta = (gpt > 0) ? std::asinh(gpz / gpt) : 0.0;
                        h_gamma_pt->Fill(gpt);
                        h_gamma_eta->Fill(geta);
                    }
                }

                // --- 4. MC Truth y resolucion ---
                if (associations) {
                    for (const auto& assoc : *associations) {
                        if (assoc.getFrom() == p) {
                            double genMass = assoc.getTo().getMass();
                            h_gen_mass->Fill(genMass);
                            h_res->Fill(recoMass - genMass);
                            break;
                        }
                    }
                }
            }
        } catch (const std::exception& e) {
            continue;
        }
    }

    std::cout << "[SUCCESS] Analysis finished." << std::endl;
    std::cout << "  - Post-fit entries:   " << h_mass->GetEntries()     << std::endl;
    std::cout << "  - Pre-fit entries:    " << h_prefit->GetEntries()   << std::endl;
    std::cout << "  - MC Truth entries:   " << h_gen_mass->GetEntries() << std::endl;
    std::cout << "  - Higgs pT entries:   " << h_pt->GetEntries()       << std::endl;
    std::cout << "  - Higgs eta entries:  " << h_eta->GetEntries()      << std::endl;

    h_mass->Write();
    h_prefit->Write();
    h_gen_mass->Write();
    h_res->Write();
    h_pt->Write();
    h_gamma_pt->Write();
    h_eta->Write();
    h_gamma_eta->Write();
    hfile->Close();
    return 0;
}

File: Analisis/make_Higgs_hists.py
import ROOT
from podio.reading import get_reader
from edm4hep import utils
p4 = utils.p4
UseEnergy = utils.UseEnergy

def main(args):
    """Main"""
    reader = get_reader(args.inputfile)
    events = reader.get("events")

    histfile = ROOT.TFile(args.outputfile, "recreate")

    # M(gg)
    Higgs_mass = ROOT.TH1D("Higgs_mass", ";M_{H} [GeV];Entries",                        100, 110.0, 140.0)
    Higgs_mass_p4 = ROOT.TH1D("Higgs_p4", ";M_{#gamma#gamma} [GeV];Entries",            100, 110.0, 140.0)
    Higgs_mass_prefit = ROOT.TH1D(
        "Higgs_mass_prefit", ";M_{#gamma#gamma} (prefit) [GeV];Entries",                 100,  90.0, 160.0
    )
    fit_delta_m = ROOT.TH1D(
        "fit_delta_m",
        ";M_{#gamma#gamma} (postfit) - M_{#gamma#gamma} (prefit) [GeV];Entries",
        100, -15.0, 15.0,
    )

    # pT
    Higgs_pt    = ROOT.TH1D("Higgs_pt",    ";p_{T}^{H} [GeV];Entries",                  100,   0.0, 200.0)
    gamma_pt    = ROOT.TH1D("gamma_pt",    ";p_{T}^{#gamma} [GeV];Entries",              100,   0.0, 150.0)

    # eta
    Higgs_eta   = ROOT.TH1D("Higgs_eta",   ";#eta^{H};Entries",                          100,  -5.0,   5.0)
    gamma_eta   = ROOT.TH1D("gamma_eta",   ";#eta^{#gamma};Entries",                     100,  -5.0,   5.0)

    for event in events:
        Higgs = event.get("Higgs_New")
        for H in Higgs:
            # Masa
            Higgs_mass.Fill(H.getMass())
            H_p4 = p4(H, UseEnergy)
            Higgs_mass_p4.Fill(H_p4.M())

            # pT y eta del Higgs
            import math
            px  = H.getMomentum().x
            py  = H.getMomentum().y
            pz  = H.getMomentum().z
            pt  = math.sqrt(px*px + py*py)
            eta = math.asinh(pz / pt) if pt > 0 else 0.0
            Higgs_pt.Fill(pt)
            Higgs_eta.Fill(eta)

            # Pre-fit y cinematica de los fotones hijos
            gamma1_p4 = p4(H.getParticles()[0], UseEnergy)
            gamma2_p4 = p4(H.getParticles()[1], UseEnergy)
            prefit_H = gamma1_p4 + gamma2_p4
            Higgs_mass_prefit.Fill(prefit_H.M())
            fit_delta_m.Fill(H_p4.M() - prefit_H.M())

            # pT y eta de cada foton
            for gamma in [H.getParticles()[0], H.getParticles()[1]]:
                gpx  = gamma.getMomentum().x
                gpy  = gamma.getMomentum().y
                gpz  = gamma.getMomentum().z
                gpt  = math.sqrt(gpx*gpx + gpy*gpy)
                geta = math.asinh(gpz / gpt) if gpt > 0 else 0.0
                gamma_pt.Fill(gpt)
                gamma_eta.Fill(geta)

    Higgs_mass.Write()
    Higgs_mass_p4.Write()
    Higgs_mass_prefit.Write()
    fit_delta_m.Write()
    Higgs_pt.Write()
    gamma_pt.Write()
    Higgs_eta.Write()
    gamma_eta.Write()
    histfile.Close()

if __name__ == "__main__":
    import argparse
    import math
    parser = argparse.ArgumentParser(
        description="Small script to make Higgs -> gg histograms via python bindings and ROOT"
    )
    parser.add_argument("inputfile", help="The input file with the data")
    parser.add_argument(
        "outputfile",
        nargs="?",
        default="Higgs_histograms.root",
        help="The output file into which the histograms will be stored",
    )
    args = parser.parse_args()
    main(args)
